include/rocksdb/file_system.h

```diff
@@ -0,0 +1,1356 @@
+                             std::shared_ptr<Logger>* result,
+                             IODebugContext* dbg) = 0;
+
+  // Get full directory name for this db.
+  virtual IOStatus GetAbsolutePath(const std::string& db_path,
+                                   const IOOptions& options,
+                                   std::string* output_path,
+                                   IODebugContext* dbg) = 0;
+
+  // OptimizeForLogRead will create a new FileOptions object that is a copy of
+  // the FileOptions in the parameters, but is optimized for reading log files.
+  virtual FileOptions OptimizeForLogRead(const FileOptions& file_options) const;
+
+  // OptimizeForManifestRead will create a new FileOptions object that is a copy
+  // of the FileOptions in the parameters, but is optimized for reading manifest
+  // files.
+  virtual FileOptions OptimizeForManifestRead(
+      const FileOptions& file_options) const;
+
+  // OptimizeForLogWrite will create a new FileOptions object that is a copy of
+  // the FileOptions in the parameters, but is optimized for writing log files.
+  // Default implementation returns the copy of the same object.
+  virtual FileOptions OptimizeForLogWrite(const FileOptions& file_options,
+                                         const DBOptions& db_options) const;
+
+  // OptimizeForManifestWrite will create a new FileOptions object that is a
+  // copy of the FileOptions in the parameters, but is optimized for writing
+  // manifest files. Default implementation returns the copy of the same
+  // object.
+  virtual FileOptions OptimizeForManifestWrite(
+      const FileOptions& file_options) const;
+
+  // OptimizeForCompactionTableWrite will create a new FileOptions object that
+  // is a copy of the FileOptions in the parameters, but is optimized for
+  // writing table files.
+  virtual FileOptions OptimizeForCompactionTableWrite(
+      const FileOptions& file_options,
+      const ImmutableDBOptions& immutable_ops) const;
+
+  // OptimizeForCompactionTableRead will create a new FileOptions object that
+  // is a copy of the FileOptions in the parameters, but is optimized for
+  // reading table files.
+  virtual FileOptions OptimizeForCompactionTableRead(
+      const FileOptions& file_options,
+      const ImmutableDBOptions& db_options) const;
+
+// This seems to clash with a macro on Windows, so #undef it here
+#ifdef GetFreeSpace
+#undef GetFreeSpace
+#endif
+
+  // Get the amount of free disk space
+  virtual IOStatus GetFreeSpace(const std::string& /*path*/,
+                                const IOOptions& /*options*/,
+                                uint64_t* /*diskfree*/,
+                                IODebugContext* /*dbg*/) {
+    return IOStatus::NotSupported();
+  }
+
+  // If you're adding methods here, remember to add them to EnvWrapper too.
+
+ private:
+  void operator=(const FileSystem&);
+};
+
+// A file abstraction for reading sequentially through a file
+class FSSequentialFile {
```

Comment:
> Why can't the existing class be changed so that the old method (by default calls the new one? If you are worried about backwards compatibility, the new one can return "not implemented" and existing implementations will work. Alternatively, you have the new methods as pure virtual and legacy FileSystems return a FsFileWrapper class.

Having the new methods as pure virtual will break the compilation of derived class implementations. Our policy is to guarantee compatibility with any code written using public interfaces. 
The only way to make the approach of using existing classes is to define the new methods as not pure virtual and have them call the old methods by default. It might save some code change in other places, but those code changes are pretty straightforward and not risky, so why not bite the bullet now and design a clean set of interfaces?

> I think an approach similar to this also eliminates the need for the "Wrapper" classes to be handled at a higher layer. Now everything can implement the new interface coming out of the Env/FileSystem and the higher-level code does not need to deal with wrappers. As existing code is changed to use the new interfaces, it should just work

The FS* wrapper classes are not really referenced outside of composite_env_wrapper.h. The higher layers are not aware of them, except for some intrusive test code.

Reply:
I agree that it would be nice to have a clean break from the existing interface but do not believe it is practical.  To do so, you are required to implement an entirely new class and create all of these wrappers to it.  It would be so much simpler (and cleaner) to just have wrapper methods in the existing classes.    In the approach of adding new classes, you are effectively deprecating the old classes (instead of just deprecating the old methods), which creates two sets of classes that will be supported and maintained.  If you add the new methods, you can (eventually) declare the old ones to be deprecated.  I believe something similar to this probably happened in the past with a MergeOperator (and FullMerge vs FullMergeV2).

And the current approach is exposing wrappers all over the place (some to the File classes, some to the FileSystem).  A cleaner approach is to hide the "dirty laundry" of the wrapping inside the classes rather than exposing it outside.

Reply:
> And the current approach is exposing wrappers all over the place (some to the File classes, some to the FileSystem). A cleaner approach is to hide the "dirty laundry" of the wrapping inside the classes rather than exposing it outside.

I frankly don't understand this statement. The wrappers are not "exposed all over the place". There are a few non-critical places where the code has access to an Env and needs to wrap a FileSystem around it. That can very well be encapsulated into a function and the caller doesn't need to know anything about it. Does that address your concern?.

Reply:
Let me re-phrase:
1.  All of the existing code in Env dealing with FileSystem is left unchanged.  Which means there are now potentially two ways (the new and the old) to do everything.  And there is no comment or other indication to show that this code is potentially going away.
-> To solve this problem, you can make the Env have a FileSystem and wrap the existing methods into that FileSystem.

2.  Because the Env does not wrap the FileSystem, there is no way to test alternative FileSystem implementations.  Most of the tests assume that they should create a wrap of the FileSystem from the Env.
-> To solve this problem, see the first point.

3.  All of the existing code dealing with the Env FileXXX classes still remains, leading to two classes that implement very similar interfaces.
-> To solve this problem, have the "old" methods call the "new" methods in the class and add a comment that these methods will be deprecated in some future release.   Create a "LegacyFileWrapper" where the "new" method calls the old one in the wrapped target class.

4.  Because the new FSxxxFile classes are not the related to the old classes, many of the tests assume the old implementation and wrap it with the new one (LegacyWrapper).  This means that if we want to implement a Custom FileSystem, it must also be a part of an Env, meaning everything is there twice.

5.  There are many other Env in the base system (HDFS, Mock, ReportFileOp, ChRoot, etc) that should be changed to a "Basic Env with a FileSystem".  Have you tried to make any of these into FileSystem-based Environments?  I am wondering if the tests would pass under such conditions.

Also, there does not seem to be any mechanism to change what the FileSystem is in the DBOptions.  You cannot override it by an option in the INI file (there is no OptionTypeInfo defined for it).

Reply:
#1 and #3 are more about documentation and communication about deprecation of old interfaces, and is orthogonal to the design choice. I don't see having two classes, one old and one new, as a big problem. The old ones will be deprecated at some point.

I see your point about tests in #2 and #4. The CompositeEnvWrapper can be used to solve this, by returning the FileSystem that it wraps around. A default implementation in Env can return a LegacyFileSystemWrapper. Does that address your concern?

> Also, there does not seem to be any mechanism to change what the FileSystem is in the DBOptions.

Can we use FileSystem::Load() in ParseDBOptions(), similar to how Env is loaded?

Reply:
I think starting with an Env class that *has* a FileSystem in it and a "GetFileSystem" could solve many of the problems.  But I think you should do at least two implementations of the FileSystem to prove that it works.

If there is a FileSystem in an Env and in the DBOptions, how do you keep them in sync?  How do you know which one will be used for what operations?

Reply:
If the Env class has a FileSystem in it, how will a user specify multiple FileSystems if they have multiple DB instances in the same process? Do we have to allocate Env objects for the sole purpose of storing a FileSystem pointer? The default Env is a singleton. How can a user use the default Env and multiple FileSystems if the two are coupled together?